graylog2-server/src/main/java/org/graylog/plugins/map/config/S3GeoIpFileService.java

```diff
@@ -0,0 +1,293 @@
+import software.amazon.awssdk.services.s3.model.S3Object;
+
+import javax.inject.Inject;
+import javax.inject.Named;
+import javax.inject.Singleton;
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.nio.file.StandardCopyOption;
+import java.time.Instant;
+
+/**
+ * Service for pulling Geo Location Processor ASN and city database files from an S3 bucket and storing them on disk.
+ * The files will initially be downloaded to a temporary location on disk, then they will be validated by the
+ * {@link org.graylog2.rest.resources.system.GeoIpResolverConfigValidator}, and after successful validation they will
+ * be moved to the active location so that the Geo Location Processor can read them. The on-disk directory location
+ * for downloaded files is S3_DOWNLOAD_LOCATION in {@link GeoIpProcessorConfig}. The file names are hardcoded to ensure
+ * that the proper files are always left active.
+ *
+ * This service is called from two places:
+ * - {@link org.graylog2.rest.resources.system.GeoIpResolverConfigValidator} will download new files when the Geo
+ * Location Processor configuration is changed and the new configuration has different S3 objects than the old.
+ * - {@link org.graylog.plugins.map.geoip.GeoIpDbFileChangeMonitorService} will check to see if new files need to be
+ * downloaded each time the service runs based on the lastModified times of the S3 objects.
+ *
+ * This class relies on the {@link DefaultCredentialsProvider} and not any settings that may be configured in the
+ * Graylog AWS plugin configuration. See https://docs.aws.amazon.com/sdk-for-java/latest/developer-guide/credentials.html#credentials-chain
+ * for how to configure your environment so that the default provider retrieves credentials properly.
+ */
+@Singleton
+public class S3GeoIpFileService {
+    private static final Logger LOG = LoggerFactory.getLogger(S3GeoIpFileService.class);
+
+    public static final String S3_BUCKET_PREFIX = "s3://";
+    public static final String ACTIVE_ASN_FILE = "asn-from-s3.mmdb";
+    public static final String ACTIVE_CITY_FILE = "standard_location-from-s3.mmdb";
+    public static final String TEMP_ASN_FILE = "temp-" + ACTIVE_ASN_FILE;
+    public static final String TEMP_CITY_FILE = "temp-" + ACTIVE_CITY_FILE;
+
+    private final S3Client s3Client;
+    private final Path asnPath;
+    private final Path cityPath;
+    private final Path tempAsnPath;
+    private final Path tempCityPath;
+
+    private Instant asnFileLastModified = Instant.EPOCH;
+    private Instant cityFileLastModified = Instant.EPOCH;
+    private Instant tempAsnFileLastModified = null;
+    private Instant tempCityFileLastModified = null;
+
+    @Inject
+    public S3GeoIpFileService(@Named(GeoIpProcessorConfig.S3_DOWNLOAD_LOCATION) String s3DownloadLocation) {
+        final S3ClientBuilder clientBuilder = S3Client.builder();
```

Comment:
When I try to start my graylog server without making any other aws related system changes, I get this error that the client can't load a default aws region. I'm not sure if this would work but perhaps we can add a default here in the java code if none are found in the system settings for customers that may never use this?

Here's the error:
```
2022-08-09 12:51:23,272 ERROR: org.graylog2.bootstrap.CmdLineTool - Startup error:
com.google.inject.CreationException: Unable to create injector, see the following errors:

1) [Guice/ErrorInjectingConstructor]: SdkClientException: Unable to load region from any of the providers in the chain DefaultAwsRegionProviderChain@399acce3: [SystemSettingsRegionProvider@570ede3b: Unable to load region from system settings. Region must be specified either via environment variable (AWS_REGION) or  system property (aws.region)., AwsProfileRegionProvider@c22cf4d: No region provided in profile: default, InstanceProfileRegionProvider@25630ede: Unable to contact EC2 metadata service.]
  at S3GeoIpFileService.<init>(S3GeoIpFileService.java:82)
  at S3GeoIpFileService.class(S3GeoIpFileService.java:62)
  at GeoIpDbFileChangeMonitorService.<init>(GeoIpDbFileChangeMonitorService.java:89)
      \_ for 5th parameter
  at GeoIpDbFileChangeMonitorService.class(GeoIpDbFileChangeMonitorService.java:67)
  while locating GeoIpDbFileChangeMonitorService
  at MapWidgetModule.configure(MapWidgetModule.java:58)
  while locating Service annotated with @Element(setName=,uniqueId=829, type=MULTIBINDER, keyType=)
```
I was able to get rid of it by adding the AWS_REGION env variable.

Reply:
Yep I noticed this during my testing today when I wiped my environment clean and started back up without any of the AWS credentials configured. The latest commit should clean it up. It leaves the detailed error message in the server log, but states that if you aren't trying to use the S3 file refresh, then you can safely ignore it. Try it out and see if that looks good to you!

Reply:
Nice! Just tested and that looks solid on my end. 

One thought: The message is kind of long and scary and will probably show up in the logs of most users. What do you think about leaving the first part at warn level, but put the "Error..." stuff at debug or info?

Reply:
I think that sounds good. INFO would still show up in the logs out of the box, so I will drop it to DEBUG.